Sprint_6

WeRiXy opened the pull request.

WeRiXy committed:
Sprint_6

pomegranate282922 commented on pom.xml:
```diff
@@ -12,6 +12,33 @@
+            <artifactId>byte-buddy</artifactId>
+            <version>1.11.16</version>
+        </dependency>
+        <dependency>
```
Зачем эта аннотация?
Нужно исправить

WeRiXy replied:
@digranate
Привет!
Из тренажера:
"Чтобы Mockito работала корректно, понадобится ещё одна библиотека — Byte Buddy. Её тоже нужно подключить:"
Удалю, конечно, если не нужна, но и из шпаргалок тогда удалите, ща все с такими пойдут похоже

pomegranate282922 replied:
Привет, я про следующую зависимость)

pomegranate282922 commented on src/test/java/com/example/CatTest.java:
```diff
@@ -0,0 +1,44 @@
+        Mockito.when(catMock.getSound()).thenReturn("Мяф");
+        Assert.assertEquals("Мяф",catMock.getSound());
+        Mockito.verify(catMock, Mockito.times(1)).getSound();
+    }
```
Нет, так тест не тестирует работу метода, а тестирует библиотеку мокито. 
иы никогда не подменяем результат того метода, который сейчас тестируем.
Нужно исправить

WeRiXy replied:
@digranate 
Из задания я понял, что каждый тест должен быть с моками) я уж не знал как изгаляться)
Я правильно понимаю, что здесь можно вообще не использовать моки, т.е. методы возвращающие результат сами, а не путем выполнения иного метода моками обставлять не нужно

pomegranate282922 replied:
Моки используем для зависимостей. Везде где есть зависимости у классов ( а это Lion и Cat) -нужны моки для зависимостей.

pomegranate282922 commented on src/test/java/com/example/CatTest.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Test
+    public void getFoodWithMock() throws Exception {
+        predatorMock = felineMock;
```
Зачем нужны дополнительные объекты? Достаточно одного мока felineMock. Нужно исправить

pomegranate282922 commented on src/test/java/com/example/CatTest.java:
```diff
@@ -0,0 +1,44 @@
+    public void getFoodWithSpy() throws Exception {
+        Cat cat = new Cat(felineSpy);
+        Assert.assertEquals(List.of("Животные", "Птицы", "Рыба"),cat.getFood());
+        Mockito.verify(felineSpy, Mockito.times(1)).eatMeat();
```
Здесь тоже можно использовать мок и стаб.
Шпионы используем только тогда, когда иначе невозможно. Здесь не тот случай.
Нужно исправить

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,63 @@
+    @Test
+    public void getFamily() {
+        String family = feline.getFamily();
+        Mockito.verify(feline, Mockito.times(1)).getFamily();
```
Здесь тестируем библиотеку мокито - нужно исправить

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,63 @@
+    public void getFamilyWithStub() throws Exception {
+        Mockito.when(feline.getFamily()).thenAnswer(invocation -> new Animal().getFamily());
+        Assert.assertEquals("Существует несколько семейств: заячьи, беличьи, мышиные, кошачьи, псовые, медвежьи, куньи",feline.getFamily());
+        Mockito.verify(feline,Mockito.times(1)).getFamily();
```
Аналогично

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,63 @@
+    @Test
+    public void getKittensNoParamInvokeGetKittensWith1() {
+        Assert.assertEquals(1, feline.getKittens());
+        Mockito.verify(feline, Mockito.times(1)).getKittens(1);
```
А вот здесь - уместно!

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,63 @@
+    @Test
+    public void testGetKittensKittensCount2Return2() {
+        Assert.assertEquals(2, feline.getKittens(2));
+        Mockito.verify(feline, Mockito.times(1)).getKittens(Mockito.anyInt());
```
А здесь опять нет)

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    public void eatMeatWitStub() throws Exception {
+        Mockito.doReturn(new ArrayList<String>()).doThrow(new RuntimeException()).when((Animal)feline).getFood(Mockito.anyString());
```
Зачем этот стаб? Что он помогает проверить? Нужно исправить

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,63 @@
+        Mockito.doReturn(new ArrayList<String>()).doThrow(new RuntimeException()).when((Animal)feline).getFood(Mockito.anyString());
+        Assert.assertEquals(new ArrayList<String>(), feline.eatMeat());
+        Mockito.verify(feline, Mockito.times(1)).getFood(Mockito.anyString());
+    }
```
Этот тест дублирует предыдущий по сути. Нужно исправить

pomegranate282922 commented on src/test/java/com/example/LionAlexTest.java:
```diff
@@ -0,0 +1,60 @@
+    @Test
+    public void getFood() throws Exception {
+        LionAlex lionAlex = Mockito.spy(new LionAlex(feline));
+        Assert.assertEquals(new ArrayList<String>(),lionAlex.getFood());
```
Разве это правильно, что метод возвращает пустой список? Нужно исправить

WeRiXy replied:
@digranate 
Подскажи еще плз как проверить, что метод getFood() вызывается именно из суперкласса и есть ли вообще смысл в этой проверки?

pomegranate282922 replied:
В смысле ты хочешь проверить, что Java работает как ожидается?) нет, это не тестируют в рамках юнит тестов. В рамках юнит тестов мы тестируем функциональность метода - что он возвращает ожидаемый результат. Зависимости мокаем, чтобы тесты были изолированными.

pomegranate282922 commented on src/test/java/com/example/LionAlexTest.java:
```diff
@@ -0,0 +1,60 @@
+    public void getFriends() throws Exception {
+        LionAlex lionAlex = Mockito.spy(new LionAlex(feline));
+        Assert.assertEquals(List.of("Марти", "Глория","Мелман"),lionAlex.getFriends());
+        Mockito.verify(lionAlex, Mockito.times(1)).getFriends();
```
Тестируем опять библиотеку мокито. Нужно исправить

pomegranate282922 commented on src/test/java/com/example/LionAlexTest.java:
```diff
@@ -0,0 +1,60 @@
+    public void doesHaveMane() throws Exception {
+        LionAlex lionAlex = Mockito.spy(new LionAlex(feline));
+        Assert.assertEquals(true,lionAlex.doesHaveMane());
+        Mockito.verify((Lion)lionAlex,Mockito.times(1)).doesHaveMane();
```
И тут

pomegranate282922 commented on src/test/java/com/example/LionTest.java:
```diff
@@ -0,0 +1,44 @@
+    Feline feline;
+
+    {
+        try {
```
А это зачем? Почему здесь?

pomegranate282922 commented on src/test/java/com/example/LionTest.java:
```diff
@@ -0,0 +1,44 @@
+    @Test
+    public void getFood() throws Exception {
+        Lion lion = new Lion(feline,"Самка");
+        Assert.assertEquals(new ArrayList<String>(),lion.getFood());
```
Здесь тоже в итоге получается странная проверка. Нужно использовать стаб.

pomegranate282922 commented on src/test/java/com/example/LionTest.java:
```diff
@@ -0,0 +1,44 @@
+@RunWith(MockitoJUnitRunner.class)
+public class LionTest {
+    @Spy
+    Lion lion;
```
Зачем это поле, если оно нигде не используется?
Нужно исправить

pomegranate282922 commented on src/test/java/com/example/LionTestParameterizedDoesHaveMane.java:
```diff
@@ -0,0 +1,50 @@
+        try {
+            Lion lion = Mockito.spy(new Lion(feline,sex));
+            Assert.assertEquals(hasMane, lion.doesHaveMane());
+            Mockito.verify(lion, Mockito.times(1)).doesHaveMane();
```
Это не нужно . Нужно исправить

pomegranate282922 commented on src/test/java/com/example/LionTestParameterizedDoesHaveMane.java:
```diff
@@ -0,0 +1,50 @@
+            Assert.assertEquals(hasMane, lion.doesHaveMane());
+            Mockito.verify(lion, Mockito.times(1)).doesHaveMane();
+        } catch (Exception e) {
+            Assert.assertEquals(hasMane, e.getMessage());
```
Так не получится.
Если самец выкинет исключение, тест будет зелёным, а это баг.
Да и использование Object некрасиво )
Нужно разделить тест - позитивные кейсы сделать через параметризацию как сейчас. А негативный - проверить отдельно в другом классе через assertThrows или параметр в аннотации Test.

WeRiXy replied:
@digranate 
Ну вот) а я так радовался, что нашел столь универсальный метод
Кстати, почему object не комильфо - я так понял тип переменной выводится из значения как при var и в параметр метода подставляется вполне определенный тип

pomegranate282922 replied:
Переменная hasMane имеет тип Object. В какой момент ты ожидаешь подстановку определённого типа?

WeRiXy replied:
Наверное, я что-то все же не то подумал)
Спасибо, что ты такая общительная на этот раз попалась мне и все быстро пояснила)

pomegranate282922 replied:
Не за что! Успехов!

pomegranate282922 reviewed (commented):
Работа хорошая! Изоляция выполнена верно ( хотя можно было ничего не менять кроме кода класса Lion).
В тестах есть вопросы.
Очень много где тестируешь не работу метода , а работу библиотеки мокито. Это неправильно и нужно везде убрать.
Соответственно, нужно очень внимательно посмотреть, где нужны шпионы, а где нет. Использовать их только там, где они реально необходимы.
Переделать тест с исключениями.
Остальное-по мелочи.

WeRiXy committed:
Sprint_6-fixed

WeRiXy committed:
Sprint_6-fixed

	modified:   src/test/java/com/example/CatTest.java
	modified:   src/test/java/com/example/LionAlexTest.java
	modified:   src/test/java/com/example/LionTest.java

WeRiXy committed:
Sprint_6-fixed

pomegranate282922 reviewed (approved):
Да, вот теперь - супер!
